pkg/queue/breaker.go

```diff
@@ -177,131 +176,137 @@ func (b *Breaker) Capacity() int {
+// state is an uint64 that has two uint32s packed into it: capacity and inFlight. The
+// former specifies how many request are allowed at any given time into the semaphore
+// while the latter refers to the currently in-flight requests.
+// Packing them both into one uint64 allows us to optimize access semantics using atomic
+// operations, which can't be guaranteed on 2 individual values.
+// The channel is merely used as a vehicle to be able to "wake up" individual goroutines
+// if capacity becomes free. It's not consistently used in accordance to actual capacity
+// but is rather a communication vehicle to ensure waiting routines are properly woken
+// up.
 type semaphore struct {
-	queue    chan struct{}
-	reducers int
-	capacity int
-	mux      sync.RWMutex
+	state atomic.Uint64
+	queue chan struct{}
 }
 
-// tryAcquire receives the token from the semaphore if there's one
-// otherwise an error is returned.
+// tryAcquire receives the token from the semaphore if there's one otherwise returns false.
 func (s *semaphore) tryAcquire() bool {
-	select {
-	case <-s.queue:
-		return true
-	default:
-		return false
+	for {
+		old := s.state.Load()
+		capacity, in := unpack(old)
+		if in >= capacity {
+			return false
+		}
+		in++
+		if s.state.CAS(old, pack(capacity, in)) {
+			return true
+		}
 	}
 }
 
-// acquire receives the token from the semaphore, potentially blocking.
+// acquire acquires capacity from the semaphore.
 func (s *semaphore) acquire(ctx context.Context) error {
-	select {
-	case <-s.queue:
-		return nil
-	case <-ctx.Done():
-		return ctx.Err()
+	for {
+		old := s.state.Load()
+		capacity, in := unpack(old)
+
+		if in >= capacity {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-s.queue:
+			}
+			// Force reload state.
+			continue
+		}
+
+		in++
+		if s.state.CAS(old, pack(capacity, in)) {
+			return nil
+		}
 	}
 }
 
-// release potentially puts the token back to the queue.
-// If the semaphore capacity was reduced in between and is not yet reflected,
-// we remove the tokens from the rotation instead of returning them back.
+// release releases capacity in the semaphore.
+// If the semaphore capacity was reduced in between and as a result inFlight is greater
+// than capacity, we don't wake up goroutines as they'd not get any capacity anyway.
 func (s *semaphore) release() error {
-	s.mux.Lock()
-	defer s.mux.Unlock()
+	for {
+		old := s.state.Load()
+		capacity, in := unpack(old)
 
-	if s.reducers > 0 {
-		s.capacity--
-		s.reducers--
-		return nil
-	}
+		if in == 0 {
+			return ErrRelease
+		}
 
-	// We want to make sure releasing a token is always non-blocking.
-	select {
-	case s.queue <- struct{}{}:
-		return nil
-	default:
-		// This only happens if release is called more often than acquire.
-		return ErrRelease
+		in--
+		if s.state.CAS(old, pack(capacity, in)) {
+			if in < capacity {
+				select {
+				case s.queue <- struct{}{}:
```

Comment:
it possibly doesn't matter (we're not often at maxCapacity!), but I'm trying to think through what happens if s.queue is at capacity here. We could be in the other loop _just_ before we try to select, but this is non-blocking and the queue is full (for a couple more milliseconds) so we'll drop this on the floor, then I think we deadlock? or no?

Reply:
I thought about guarding the entire semaphore with a capacity check that would ensure we never exceed int32Max. In that case, an overflow here is, I believe, a degenerate case in that you're holding it wrong because there should never be more wakeups than capacity.

Reply:
yah, in practice I don't think this is a bug worry since you're never in practice going to have a capacity anywhere near MaxInt32 so this would never really happen (maybe a little check somewhere is nice as a guard). Out of interest, do we actually need to do a non-blocking send here at all then? We can never be at capacity so the buffer is always big enough so we're never going to block (I think?) -- do the benchmarks suffer if you drop the select?

Reply:
I haven't actually tried, it's taken from the previous implementation. Maybe we should return an error as we did before. The error mode of blocking here is just super bad :sweat_smile:

Reply:
I think an error seems legit, since this is a pretty exceptional case and one we'll probably struggle to debug so the more clues the better (I almost want to panic, tbh, but an error will do).

Reply:
I actually stand corrected: When in doubt, we generate more wakeups than we might need. `UpdateCapacity` for instance generates a wakeup for each bit of capacity added, regardless of there being a waiter or not. The same goes for `Release`. With the data available, we can't really synchronize the wakeup generation with the waiters (which is kind of the whole point here) and thus we might at some point hit the ceiling here, which wouldn't be an issue I think because we generate more wakeups than we need anyway but we'd never generate too little wakeups.

Reply:
Added a comment to that extent.

Reply:
interesting! ok, I think this only comes in to play when we're at waaaaaay higher capacity than can ever really be the case functionally anyway, but the comment does make me feel better